### PR TITLE
Add Base64 Encoding/Decoding and Enhanced String Handling for Bytes

### PR DESCRIPTION
**Pull request by Shigoto-dev19**
## Description

All changes revolve around the provable **Bytes** class.

## Changes

- Added `base64Encode` and `base64Decode` methods.
- Added unit tests for the new Base64 methods.

### Notes

- For detailed instructions on using Base64 encode/decode in o1js and an overview of provable Base64 constraints, refer to the `o1js-base64` [repo README](https://github.com/Shigoto-dev19/o1js-base64?tab=readme-ov-file#o1js-base64-encodedecode).
- For an example of zkPrograms using Base64, see this [file](https://github.com/Shigoto-dev19/o1js-base64/blob/main/src/run.ts).

### CODE REVIEW

**Review by mitschabaude — CHANGES_REQUESTED** [review]
Thanks @Shigoto-dev19! This is an incredible addition to o1js.

I'll review in detail later but have some quick feedback: You can't remove methods like `toHex()` and `fromHex()` because this is a breaking change and would move o1js to version 2.0 according to semver, which we don't want.

**Review by mitschabaude — COMMENTED** [review]
I reviewed it, have a few more comments, looks great overall

**Comment by mitschabaude** [comment]
note that the tests are just failing because you changed the `bindings` and `mina` submodules

**Comment by Shigoto-dev19** [comment]
> I reviewed it, have a few more comments, looks great overall

Thank you for the review! I'll address the comments and update the branch :)

File: src/lib/provable/bytes.ts [hunk 1]
```diff
@@ -132,3 +247,113 @@ function createBytes(size: number): typeof Bytes {
...
 }
+
+/**
+ * Decodes a Base64 character to its original value.
+ * Adapted from the algorithm described in: http://0x80.pl/notesen/2016-01-17-sse-base64-decoding.html#vector-lookup-base
+ *
+ * @param input - The Base64 encoded byte to be decoded.
+ * @returns - The corresponding decoded value as a Field.
+ */
+function base64DLookup(input: UInt8): Field {
```
[the remark this answers is not in the document]
Comment (Shigoto-dev19) [reply src/lib/provable/bytes.ts:258]: I agree!

File: src/lib/provable/bytes.ts [hunk 2]
```diff
@@ -60,12 +62,111 @@ class Bytes {
...
+  base64Decode(byteLength: number): Bytes {
+    const encodedB64Bytes = this.bytes;
+
+    const charLength = encodedB64Bytes.length;
+    assert(
+      charLength % 4 === 0,
+      'Input base64 byte length should be a multiple of 4!'
+    );
+
+    let decodedB64Bytes: UInt8[] = [];
```
[the remark this answers is not in the document]
Comment (Shigoto-dev19) [reply src/lib/provable/bytes.ts:118]: Awesome catch!

File: CHANGELOG.md [hunk 3]
```diff
@@ -17,6 +17,15 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
...
 ## [Unreleased](https://github.com/o1-labs/o1js/compare/6a1012162...HEAD)
 
+### Added
+
+- Added `base64Encode()` and `base64Decode(byteLength)` methods to the **Bytes** type class.
+- Added `toString(encoding?: BufferEncoding, start?: number, end?: number)` method for converting Bytes to string with different encodings and optional byte range selection.
+
+### Changed
+
+- Changed fromString(s) to support more encodings, aligned with the native BufferEncoding union type.
```
[the remark this answers is not in the document]
Comment (Shigoto-dev19) [reply CHANGELOG.md:27]: I removed the changed subsection as `Buffer` is no longer used.

File: CHANGELOG.md [hunk 4]
```diff
@@ -17,6 +17,15 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ## [Unreleased](https://github.com/o1-labs/o1js/compare/6a1012162...HEAD)
 
+### Added
+
+- Added `base64Encode()` and `base64Decode(byteLength)` methods to the **Bytes** type class.
+- Added `toString(encoding?: BufferEncoding, start?: number, end?: number)` method for converting Bytes to string with different encodings and optional byte range selection.
```
[the remark this answers is not in the document]
Comment (Shigoto-dev19) [reply CHANGELOG.md:23]: I deleted the `Bytes.toString()` both as a method and in changelog

**Review by mitschabaude — APPROVED** [review]

**Pull request merged**